### ISSUE TITLE
arcswap: better ensure imbalance thresholds are not exceeded

### ISSUE THREAD

**Issue by hhirtz** [LIHPC-Computational-Geometry/coupe#197]
Follow up of #193

**Comment by hhirtz** [comment]
![histo](https://user-images.githubusercontent.com/51088794/187863889-6f00329d-40be-4bfd-ae4b-bb72f11061f3.svg)

**Comment by SebastienMorais** [comment]
[reproducer_data.zip](https://github.com/LIHPC-Computational-Geometry/coupe/files/9493983/reproducer_data.zip)

This is a reproducer that even fails for 1 thread. Sequence of calls :

```
mesh-part -v -a rcb,7 -a vn-best -a arcswap,0.05 --mesh triangles.mesh --weights triangles.spike.weights > triangles.spike.part
part-info -m triangles.mesh -w triangles.spike.weights -p triangles.spike.part
```

**Comment by cedricchevalier19** [comment]
@SebastienMorais Can you write the expected output of the second command ?

**Comment by cedricchevalier19** [comment]
I think I know understand the bug: main loop is using [fold](https://docs.rs/rayon/latest/rayon/iter/trait.ParallelIterator.html#method.fold) assuming it will produce the result will have the same length than the number of rayon threads. It is not the case, so the `max_part_weight` upper limit is not tight enough.

```rust
        partition
            .par_iter()
            .enumerate()
            .fold(
                || (Metadata::default(), Vec::new(), part_weights.clone()),
                |(mut metadata, mut cut, mut part_weights), (vertex, initial_part)| { ... },
            )
            .map(|(metadata, _, _)| { ... })
```

#204 does not seem to fix this issue, with 1 thread I still have 4 iterations in the `map` after the `fold`.

Moreover, I am really not convinced by how the method is currently implemented, especially why looping on the partition is enough.

**Comment by hhirtz** [comment]
yes this is the bug. Thanks for the 1-thread tests, I closed #204 since it does not fix the issue.

> Moreover, I am really not convinced by how the method is currently implemented, especially why looping on the partition is enough.

This is indeed not enough and the loop needs to be replaced. As you can see on the graph above, applying arcswap twice on a random partition produces better results than applying it once.

**Comment by hhirtz** [comment]
I cleaned up the script used to generate [the above graph][0] and made a PR here: #206 

[0]: https://github.com/LIHPC-Computational-Geometry/coupe/issues/197#issuecomment-1233899661
